Discussion part 4: comments, nested comments/replies + ability to comment on discussion and comments

darrenvong opened the pull request:
<!---
Please describe the pull request below.
If it closes an issue, make sure to write "closes #1234"
If there is an issue but it isn't completely closed, still refer to the issue number, eg. "part of #1234"
--->
Covers more of #543.

Discussion feature as described in PR title.

I'll add screenshots later. Tests left to do but feel free to do an early review if you got time

<!---
Checklists - you can remove one that is not applicable (ie. remove backend checklist if you only worked on frontend)
If you need help with any of these, please ask :)
--->
**Frontend checklist**
- [x] Formatted my code with `yarn format && yarn lint --fix`
- [x] There are no warnings from `yarn lint`
- [x] There are no console warnings when running the app
- [x] Added any new components to storybook
- [x] Added tests where relevant
- [x] All tests pass
- [x] Clicked around my changes running locally and it works
- [x] Checked Desktop, Mobile and Tablet screen sizes


<!---
Remember to request review from couchers-org/frontend, couchers-org/@backend or an individual.
Once your code is approved, remember to merge it if you have write access
--->

darrenvong committed:
feat: rename discussion folder -> discussions

darrenvong committed:
feat: move discussion related components to discussions folder

darrenvong committed:
feat: WIP reply tree/reply components

darrenvong committed:
feat: move thread data fetching into RepliesTree

darrenvong committed:
feat: refactor comment into own component

darrenvong committed:
feat: use markdown editor to create discussion

darrenvong committed:
feat: Add comment/reply form to top level comments

darrenvong committed:
feat: fix getThread mock in Storybook/tests

Without the fix, the naive getThread mock causes the comment to nest
infinitely, as each comment will get the same reply list, and that in
turn will have more comments and so on, resulting in infinite rendering
loop

darrenvong committed:
feat: add comment form to discussion

darrenvong committed:
feat: stop comment form from closing if default is to keep it opened

darrenvong committed:
feat: simplify comment card grid styling

lucaslcode commented:
With the URLs, looking at `docs/URLs.py` what you have now is correct. But the community (aka official cluster) id is in the database, so it should be trivial to include it in `GetDiscussion`, so in that case I think we can put the discussion page inside the community. Dunno if you want to do it in this PR or not?

darrenvong commented:
How would this work with the URL matching though, since the community pages start with `/community` and a single discussion starts with `/discussion`? I'm guessing we'll probably have to duplicate or split the common bits into a shared component so both community and discussion can use it?

Also how would things look roughly layout wise? Instead of the tabs with the different things, we replace it with the discussion in the same place if we are on a single discussion?

lucaslcode commented:
Yes, I guess the community page would need some refactoring so it can take a child?
The top bit of the community page (everything up to line 190) can be `CommunityBase.tsx`, and it can optionally take a `communityId` prop - if it doesn't have it, it gets the id from the url. And then put `{ children }` just below it.

Then the real community page can have `<CommunityBase>/*<Switch>s here*/</CommunityBase>` and discussions page can be `<CommunityBase communityId={discussion.communityId}>/*discussion page*/</CommunityBase>`

lucaslcode commented:
Oh and I didn't answer your question - I think what would make the most sense is the tab bar is still there, just none of them are selected

darrenvong committed:
feat: remove redundant component from merge conflict horror

darrenvong committed:
feat: make comment form closeable if its default state is hidden

darrenvong committed:
feat: fix more bugs introduced from the merge conflict

darrenvong committed:
feat: simplify tab bar

darrenvong commented:
Hmm I think I'll do the page refactor in another PR, as I imagine this one will get pretty huge otherwise with big diffs as is usually the case when refactoring/shifting things around

darrenvong committed:
feat: add tests for comment tree and commenting functionality

lucaslcode commented:
Didn't notice last time - the cards on the community page are slightly broken on mobile: 
![image](https://user-images.githubusercontent.com/32044095/117859036-6b455600-b286-11eb-8237-ed4198b53a5e.png)

lucaslcode commented:
~~I think the button to add a nested comment should move outside the card and at the bottom of any other nested comments:~~ 
![image](https://user-images.githubusercontent.com/32044095/117859555-0fc79800-b287-11eb-87c3-27066a2c4ecd.png)
~~Otherwise, it's not intuitive what happens when there are already nested comments (like in this image).~~

I see, nested comments are also latest first. Hmm. ~~I wonder if they should be oldest first? That might require backend changes.~~

After thinking about it, how about making the latest comments at the bottom, and displaying a "load earlier comments" link at the top? And indeed moving the button as I have in the image.

There also seems to be a bit too much bottom padding in the cards

lucaslcode commented on app/frontend/src/features/communities/discussions/Comment.tsx:
```diff
@@ -0,0 +1,122 @@
+        data-testid={COMMENT_TEST_ID}
+        key={comment.createdTime!.seconds}
+      >
+        <Avatar user={user} className={classes.avatar} isProfileLink={false} />
```
```suggestion
        <Avatar user={user} className={classes.avatar} />
```

lucaslcode commented on app/frontend/src/features/communities/discussions/CommentForm.tsx:
```diff
@@ -0,0 +1,114 @@
+    ({ content }) => service.threads.postReply(threadId, content),
+    {
+      onSuccess() {
+        queryClient.invalidateQueries(threadKey(threadId));
```
I can't remember what you decided last time, but this is suffering from the same problem where on a slow connection, the loading state finishes when the new comment is posted but before it actually appears.

darrenvong replied:
We went with showing a refetch loading state instead, since optimistically updating the data in an infinite query is much more of a pain than a normal query. There is one in `Comment`, but I may have to adjust the position of it to be at the bottom if we are going with having the comment input near the last comment.

lucaslcode replied:
I see it now in storybook. For some reason I didn't before, maybe I just missed it.

lucaslcode commented on app/frontend/src/features/communities/discussions/Comment.tsx:
```diff
@@ -0,0 +1,122 @@
+      marginBlockStart: theme.spacing(2),
+    },
+    marginBlockStart: theme.spacing(2),
+    marginInlineStart: theme.spacing(5),
```
How about
```suggestion
    marginInlineStart: theme.spacing(3),
    marginInlineStart: `clamp(${theme.spacing(2)}, 5vw, ${theme.spacing(5)})`,
```
Mobile is most likely not to support clamp, so I made the fallback a bit less

darrenvong replied:
That's a nice trick actually rather than needing media queries. Still had to do an annoying workaround with '&' to select the same thing again and put in the clamp since you can't have duplicate object keys :(

lucaslcode commented on app/frontend/src/features/communities/discussions/DiscussionPage.tsx:
```diff
@@ -21,24 +21,22 @@ import { CREATED_AT, PREVIOUS_PAGE, UNKNOWN_USER } from "../constants";
+  root: {
+    paddingBlockEnd: theme.spacing(5),
+    "& .tui-editor-contents": {
+      marginBottom: theme.spacing(3),
```
This should be in the relevant components (or more specific) rather than here. It's appropriate for the discussion content, but it's also affecting the cards, where it's too much margin (imo)

lucaslcode commented on app/frontend/src/features/communities/discussions/DiscussionPage.tsx:
```diff
@@ -96,8 +94,13 @@ export default function DiscussionPage() {
+              <Avatar
+                user={discussionCreator}
+                className={classes.avatar}
+                isProfileLink={false}
```
remove

darrenvong replied:
I blame the brutal merge conflict with too many changes 😤   but good catch 😄

lucaslcode reviewed (changes requested):
Haven't looked through all the code yet - these are just some design comments

darrenvong commented:
> I think the button to add a nested comment should move outside the card and at the bottom

Basing off Reddit and Facebook, they all have the reply button right under/near the comment you are replying to, so not too sure about putting it under the last nested comment/reply. That could be unexpected behaviour for users who are used to other sites. What's your main objection on keeping the button inside the comment card in that case?

> how about making the latest comments at the bottom, and displaying a "load earlier comments" link at the top?

I know we did for chat, but since these comments also has a reply button (i.e. something keyboard focusable), I'm not sure it's a good idea to do the same CSS trick of reversing the display order of the comments as keyboard users would probably end up going from bottom to top if they tab between buttons. Getting that to work "out of the box" as we do now might be a bit of a pain.

Reversing the data is a potential option, but feels like extra work that doesn't need to be done by the frontend if the backend gives it in the right order to start with. And the operation is linear time so it's gonna make our page slower as it grows (although unlikely to be an immediate problem for some time) so probably better done in the backend.

lucaslcode commented:
The difference is, reddit has infinite reply levels, and clicking reply adds another level. Here, it will only go one level deep - I think the difference is important.

Facebook does have the reply button on the comment, but when you click it, it opens the comment box underneath all the other comments. We could do that!

> keyboard users would probably end up going from bottom to top

Good point. Let's reverse it in the backend then - changing `.desc()` to `.asc()` in `threads.py` lines 67 and 90 should be enough, but it might break some tests, so leave it to another PR if you want. But treat it as if it has been done?

lucaslcode commented on app/frontend/src/features/communities/discussions/CommentForm.tsx:
```diff
@@ -0,0 +1,114 @@
+}
+
+export default function CommentForm({
+  shown = false,
```
Would this be better as `alwaysShow` or `toggleable`?

lucaslcode commented on app/frontend/src/features/communities/discussions/CommentTree.tsx:
```diff
@@ -57,52 +53,39 @@ export default function CommentTree({ threadId }: CommentTreeProps) {
-                  </div>
-                </Card>
+                <Comment
+                  key={comment.createdTime?.seconds}
```
Why not `threadId`? There could easily be comments posted in the same second.

darrenvong replied:
That's a good point, `threadId` is much better - I remember I was scrambling to find something unique but probably overlooked this with the whole thread/comment confusion at the beginning 😅

darrenvong committed:
feat: fix discussion card comment count clipping off on small screens

darrenvong commented:
> Good point. Let's reverse it in the backend then - changing `.desc()` to `.asc()` in `threads.py` lines 67 and 90 should be enough, but it might break some tests, so leave it to another PR if you want. But treat it as if it has been done?

Wouldn't that break the pagination logic though, as the limit bit might be dependent on it? I think I'll leave this to the backend folks for that one... but yeah happy to treat it as done for the sake of this PR

>  it opens the comment box underneath all the other comments. We could do that!

I'll have a play around with that to see how much more complex that might make it than just placing it near the bottom and probably go with whatever's simpler. The extra complexity being needing to keep refs around probably to scroll to the comment box, but it might not be too bad

darrenvong committed:
feat: address discussion/comment styling feedback

aapeliv commented:
> > Good point. Let's reverse it in the backend then - changing `.desc()` to `.asc()` in `threads.py` lines 67 and 90 should be enough, but it might break some tests, so leave it to another PR if you want. But treat it as if it has been done?
> 
> Wouldn't that break the pagination logic though, as the limit bit might be dependent on it? I think I'll leave this to the backend folks for that one... but yeah happy to treat it as done for the sake of this PR

See #1223

darrenvong committed:
feat: move reply input below the last reply

darrenvong committed:
feat: move refetch loading state below last reply + update tests

darrenvong commented:
I think I addressed all the comments now apart from needing to add the "load earlier comments" button somewhere. Feel free to have another look through in between

darrenvong committed:
feat: improve scroll behaviour to comment box in Safari

Even with the polyfill, Safari doesn't support the "block: center"
option of smooth scrolling, which nicely scrolls the input into centre
of the screen.

However, there's a Webkit proprietary "scrollIntoViewIfNeeded" which
when called with true, it does scroll the input to centre of screen.

I think this is a worthy tradeoff on Safari than keeping smooth
scrolling, which always smooth scroll the input to the top of screen
and get clipped off by our fixed position nav bar.

darrenvong committed:
feat: Add load more comments button to comment tree and comment

darrenvong commented:
Ok, list of messages is now reversed in the frontend, and the load more button is done now.

If you manage to check before I have more time to add Storybook (need to hack the mock a bit to deal with pagination), you'll probably have to pull this down locally and change the `pageSize` inside getThread to something smaller temporarily to see the button, but otherwise the functionality is pretty much there now

darrenvong commented on app/frontend/src/features/communities/discussions/Comment.tsx:
```diff
@@ -0,0 +1,180 @@
+  useEffect(() => {
+    if (showCommentForm && commentFormRef.current) {
+      if (isSafari()) {
+        commentFormRef.current.scrollIntoViewIfNeeded(true);
```
Safari doesn't support smooth scrolling AND scrolling element to centre of the screen even with the polyfill, but calling `scrollIntoViewIfNeeded(true)` does it but without smooth scrolling. I thought for Safari only, it's worth giving up smooth scrolling for this, otherwise the input always scroll to top (default option) of screen and so the input gets clipped under the nav bar, which is probably more annoying than not having smooth scrolling imo.

darrenvong replied:
Also since I'm not measuring anything and probably because of the 1-2s delay caused by the animation revealing the input, I managed to get away with using `useEffect` to do this rather than with `useLayoutEffect` :D

lucaslcode replied:
Maybe we need a better polyfill - does this one work? https://github.com/magic-akari/seamless-scroll-polyfill

darrenvong replied:
Haven't tried yet but from looking at the code, it looks more promising and comprehensive than the original polyfill we are using now (which this is a fork of!)

darrenvong replied:
That other polyfill worked like a charm, nice find!

darrenvong committed:
feat: fix broken tests due to reversed message list change

darrenvong committed:
feat: replace smoothscroll-polyfill with seamless-scroll-polyfill

The latter is a fork of the former, but has complete support for all of
the scroll into view options

darrenvong committed:
feat: update Storybook mock to show pagination buttons in discussion

darrenvong committed:
feat: add tests for pagination logic in UI

lucaslcode commented on app/frontend/src/features/communities/discussions/Comment.tsx:
```diff
@@ -0,0 +1,175 @@
+                .map((reply) => {
+                  return <Comment key={reply.threadId} comment={reply} />;
+                })}
+              {!showLoadMoreButton && isCommentsRefetching && (
+                <CircularProgress />
+              )}
```
Aha! I wasn't going crazy - this doesn't appear for the first nested comment, because it is inside `hasAtLeastOnePage`

lucaslcode commented on app/frontend/package.json:
```diff
@@ -23,7 +23,7 @@
     "react-query": "^3.15.2",
     "react-router-dom": "^5.2.0",
     "react-scripts": "4.0.3",
-    "smoothscroll-polyfill": "^0.4.4",
```
also remove the types dependency

lucaslcode commented on app/frontend/src/features/communities/discussions/Comment.tsx:
```diff
@@ -0,0 +1,175 @@
+              {!showLoadMoreButton && isCommentsRefetching && (
+                <CircularProgress />
+              )}
+              {showLoadMoreButton && (
+                <Button
+                  className={classes.loadMoreRepliesButton}
+                  loading={isFetchingNextPage}
+                  onClick={() => fetchNextPage()}
+                >
+                  {LOAD_MORE_REPLIES}
+                </Button>
```
I was imagining this would also be a "Load earlier replies" button, but maybe it's not so important

lucaslcode reviewed (changes requested):
It's great, just a few very minor things

darrenvong committed:
feat: fix refetch loading state not showing

Edge case on when there are no replies to start with since the loading
state was placed to only show when there's at least one reply

lucaslcode commented on app/frontend/src/features/communities/discussions/Comment.tsx:
```diff
@@ -145,16 +149,13 @@ export default function Comment({ topLevel = false, comment }: CommentProps) {
-                <CircularProgress />
-              )}
               {showLoadMoreButton && (
                 <Button
```
Can this also be moved to above the list of replies?

lucaslcode reviewed (changes requested):
Last thing

darrenvong committed:
feat: move load earlier replies button above first reply

darrenvong committed:
feat: fix tests that were flaky with appropriate mock/longer timeout

lucaslcode reviewed (approved):
lgtm pending checks!
